File: r/src/R/ai/h2o/sparkling/ml/models/H2OMOJOModel.R
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

source(file.path("R", "H2OMOJOModelBase.R"))

H2OMOJOModel.createFromMojo <- function(pathToMojo, settings = H2OMOJOSettings.default()) {
  sc <- spark_connection_find()[[1]]
  jmojo <- invoke_static(sc, "ai.h2o.sparkling.ml.models.H2OMOJOModel", "createFromMojo", pathToMojo, settings$toJavaObject())
  className <- invoke(invoke(jmojo, "getClass"), "getSimpleName")
  if (className == "H2OTreeBasedSupervisedMOJOModel") {
    H2OTreeBasedSupervisedMOJOModel(jmojo)
  } else if (className == "H2OTreeBasedUnsupervisedMOJOModel") {
    H2OTreeBasedUnsupervisedMOJOModel(jmojo)
  } else if (className == "H2OSupervisedMOJOModel") {
    H2OSupervisedMOJOModel(jmojo)
  } else if (className == "H2OUnsupervisedMOJOModel") {
    H2OUnsupervisedMOJOModel(jmojo)
  } else if (className == "H2OFeatureMOJOModel") {
    H2OFeatureMOJOModel(jmojo)
  } else {
    H2OAlgorithmMOJOModel(jmojo)
  }
}

#' @export H2OMOJOModel
H2OMOJOModel <- setRefClass("H2OMOJOModel", methods = list(
  getModelDetails = function() {
    invoke(.self$jmojo, "getModelDetails")
  },
  getDomainValues = function() {
    invoke(.self$jmojo, "getDomainValues")
  },
  getTrainingMetrics = function() {
    invoke(.self$jmojo, "getTrainingMetrics")
  },
  getValidationMetrics = function() {
    invoke(.self$jmojo, "getValidationMetrics")
  },
  getCrossValidationMetrics = function() {
    invoke(.self$jmojo, "getCrossValidationMetrics")
  },
  getCrossValidationMetricsSummary = function() {
    outputFrame <- invoke(.self$jmojo, "getCrossValidationMetricsSummary")
    sdf_register(outputFrame)
  },
  getCurrentMetrics = function() {
    invoke(.self$jmojo, "getCurrentMetrics")
  },
  getTrainingParams = function() {
    invoke(.self$jmojo, "getTrainingParams")
  },
  getModelCategory = function() {
    invoke(.self$jmojo, "getModelCategory")
  },
  getScoringHistory = function() {
    outputFrame <- invoke(.self$jmojo, "getScoringHistory")
    sdf_register(outputFrame)
  },
  getFeatureImportances = function() {
    outputFrame <- invoke(.self$jmojo, "getFeatureImportances")
    sdf_register(outputFrame)
  }
))

#' @export H2OFeatureMOJOModel
H2OFeatureMOJOModel <- setRefClass("H2OFeatureMOJOModel", contains = c("H2OMOJOModel", "H2OMOJOModelBase"))

#' @export H2OAlgorithmMOJOModel
H2OAlgorithmMOJOModel <- setRefClass("H2OAlgorithmMOJOModel", contains = c("H2OMOJOModel", "H2OAlgorithmMOJOModelBase"))

#' @export H2OSupervisedMOJOModel
H2OSupervisedMOJOModel <- setRefClass("H2OSupervisedMOJOModel", contains = ("H2OAlgorithmMOJOModel"), methods = list(
  getOffsetCol = function() {
    invoke(.self$jmojo, "getOffsetCol")
  }
))

#' @export H2OTreeBasedSupervisedMOJOModel
H2OTreeBasedSupervisedMOJOModel <- setRefClass("H2OTreeBasedSupervisedMOJOModel", contains = ("H2OSupervisedMOJOModel"), methods = list(
  getNtrees = function() {
    invoke(.self$jmojo, "getNtrees")
  }
))

#' @export H2OUnsupervisedMOJOModel
H2OUnsupervisedMOJOModel <- setRefClass("H2OUnsupervisedMOJOModel", contains = ("H2OMOJOModel"), fields = list(jmojo = "ANY"), methods = list(
))

#' @export H2OTreeBasedUnsupervisedMOJOModel
H2OTreeBasedUnsupervisedMOJOModel <- setRefClass("H2OTreeBasedUnsupervisedMOJOModel", contains = ("H2OUnsupervisedMOJOModel"), methods = list(
  getNtrees = function() {
    invoke(.self$jmojo, "getNtrees")
  }
))

File: py/src/ai/h2o/sparkling/ml/models/H2OMOJOModel.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from pyspark.ml.param import *
from pyspark.ml.util import _jvm

from ai.h2o.sparkling.Initializer import Initializer
from ai.h2o.sparkling.ml.models import H2OMOJOSettings
from ai.h2o.sparkling.ml.models.H2OMOJOModelBase import H2OMOJOModelBase
from ai.h2o.sparkling.ml.params.H2OTypeConverters import H2OTypeConverters


class H2OMOJOModel(H2OMOJOModelBase):

    @staticmethod
    def createFromMojo(pathToMojo, settings=H2OMOJOSettings.default()):
        # We need to make sure that Sparkling Water classes are available on the Spark driver and executor paths
        Initializer.load_sparkling_jar()
        javaModel = _jvm().ai.h2o.sparkling.ml.models.H2OMOJOModel.createFromMojo(pathToMojo, settings.toJavaObject())
        className = javaModel.getClass().getSimpleName()
        if className == "H2OTreeBasedSupervisedMOJOModel":
            return H2OTreeBasedSupervisedMOJOModel(javaModel)
        elif className == "H2OSupervisedMOJOModel":
            return H2OSupervisedMOJOModel(javaModel)
        elif className == "H2OUnsupervisedMOJOModel":
            return H2OUnsupervisedMOJOModel(javaModel)
        else:
            return H2OMOJOModel(javaModel)

    def getModelDetails(self):
        return self._java_obj.getModelDetails()

    def getDomainValues(self):
        return H2OTypeConverters.scalaMapStringDictStringToStringDictString(self._java_obj.getDomainValues())

    def getTrainingMetrics(self):
        return H2OTypeConverters.scalaMapStringStringToDictStringAny(self._java_obj.getTrainingMetrics())

    def getValidationMetrics(self):
        return H2OTypeConverters.scalaMapStringStringToDictStringAny(self._java_obj.getValidationMetrics())

    def getCrossValidationMetrics(self):
        return H2OTypeConverters.scalaMapStringStringToDictStringAny(self._java_obj.getCrossValidationMetrics())

    def getCurrentMetrics(self):
        return H2OTypeConverters.scalaMapStringStringToDictStringAny(self._java_obj.getCurrentMetrics())

    def getTrainingParams(self):
        return H2OTypeConverters.scalaMapStringStringToDictStringAny(self._java_obj.getTrainingParams())

    def getModelCategory(self):
        return self._java_obj.getModelCategory()


class H2OSupervisedMOJOModel(H2OMOJOModel):

    def getOffsetCol(self):
        return self._java_obj.getOffsetCol()


class H2OTreeBasedSupervisedMOJOModel(H2OSupervisedMOJOModel):

    def getNtrees(self):
        return self._java_obj.getNtrees()


class H2OUnsupervisedMOJOModel(H2OMOJOModel):
    pass
